caluma/caluma_form/validators.py

```diff
@@ -285,67 +358,55 @@ def _validate_required(self, validation_context):  # noqa: C901
         question slugs.
         """
         required_but_empty = []
-        visible_questions = []
 
         answers = validation_context["answers"]
 
         q_jexl = jexl.QuestionJexl(validation_context)
         for question in validation_context["questions"].values():
             try:
-                expr = "is_hidden"
-                is_hidden = q_jexl.is_hidden(question)
-
-                if not is_hidden:
-                    visible_questions.append(question.slug)
-                    expr = "is_required"
-
-                    is_required = q_jexl.is_required(question)
-                    if is_required and answers.get(question.slug) in EMPTY_VALUES:
-                        required_but_empty.append(question.slug)
-
-                    if question.type == Question.TYPE_FORM:
-                        # form questions's answers are still in the top level document
+                if question.slug not in validation_context["visible_questions"]:
+                    continue
+                is_required = q_jexl.is_required(question)
+                if is_required and answers.get(question.slug) in EMPTY_VALUES:
+                    required_but_empty.append(question.slug)
+
+                if question.type == Question.TYPE_FORM:
+                    # form questions's answers are still in the top level document
+                    sub_context = {
+                        **validation_context,
+                        "questions": {
+                            q.slug: q for q in question.sub_form.questions.all()
+                        },
+                    }
+                    self._validate_required(sub_context)
+                elif question.type == Question.TYPE_TABLE and is_required:
```

Comment:
Just a question: Why do we validate TableQuestions only if `is_required`, but FormQuestions are validates anyway?

Reply:
I think FormQuestions don't have a real required state on their own - it's given by the requiredness of the questions the form contains.

Reply:
Ah makes sense, thanks